src/spec/response.ts

```diff
@@ -1,50 +1,80 @@
 
+// Matches an attachment wildcard (e.g. <IMAGE>) at the _end_ of responseData.
+//
+// The matched part is just the word in the brackets ('IMAGE'), which can be
+// directly compared against the Attachment enum (from which it was generated).
+const attachmentAlternatives = Object.values(Attachment).join('|');
+const attachmentsRegex = new RegExp(`\\s*<(${attachmentAlternatives})>$`);
+
 export class Response {
-  responseType: MessageType;
+  attachments: Attachment[];
   private textMatchChecker: RegExp;
+  private bodyText: string;
   original: string;
 
   constructor(responseData: string) {
+    this.attachments = [];
     this.original = responseData.trim();
-    switch (this.original) {
-      case '<IMAGE>':
-        this.responseType = MessageType.Image;
-        break;
-      case '<CARDS>':
-        this.responseType = MessageType.Card;
-        break;
-      default:
-        this.responseType = MessageType.Text;
-        this.textMatchChecker = new RegExp(Response.transformTags(this.original));
+    this.bodyText = this.original;
+
+    // Construct a list of attachments from the attachment wildcards at the
+    // end of the response data.  The end result is that `this.bodyText`
+    // contains only the body text and potentially some inline wildcards;
+    // and `this.attachments` contains a list of attachment types, in the
+    // same order as they originally were in the supplied data.
+    let match;
+    while ((match = this.bodyText.match(attachmentsRegex)) != null) {
+      this.attachments.push(match[1]);
+      this.bodyText = this.bodyText.slice(0, match.index);
     }
+    this.attachments.reverse();
+
+    // Construct a regex to match the remaining body text in terms of its
+    // inline wildcards.
+    this.textMatchChecker = new RegExp(Response.transformTags(this.bodyText));
   }
 
   matches(message: Message): boolean {
-    if (!message.messageTypes.includes(this.responseType)) {
+
+    // If we are expecting a message body, check that we have one...
+    if (this.bodyText && !message.text) {
       return false;
-    } else if (this.responseType === MessageType.Text) {
-      return message.text.trim().match(this.textMatchChecker) !== null;
-    } else {
-      return true;
     }
+
+    // ...and check that the content is the same.
+    //
+    // Note that we ignore the message body altogether if the spec doesn't
+    // include one.  This is for backwards compatibility with when <IMAGE>
+    // and <CARDS> could only be used in place of the entire message.
+    if (this.bodyText && message.text.trim().match(this.textMatchChecker) === null) {
+      return false;
+    }
+
+    // Check that we have the same number, kind, and order of attachments.
+    // We can use JSON.stringify here since we know attachments is just a
+    // simple array of strings.
+
+    // Check that we have the right number of attachments.
+    if (JSON.stringify(this.attachments) !== JSON.stringify(message.attachments)) {
```

Comment:
Stringify has this tricky behavior where ordering of keys is only guaranteed in Node environment **IF** the object is the same (same pointer). Even if two objects look the same, they will not necessarily get stringified in the same way. 

Use https://github.com/substack/json-stable-stringify to do a sort before the stringify.

Reply:
Interesting. I assumed that stringifying an Array<String> would be safe. If you like I can just revert ac0dd3b to avoid bringing in another library?

Reply:
Yeah that sounds like a good idea